### PR TITLE
NeuralMachineTranslation Model and Text2Sparql Example

### PR DESCRIPTION
**Pull request by vinchg**
* added NeuralMachineTranslation model which wraps HF's seq2seq models
* added NMT example with MeetKai's Text2Sparql data with training and evaluation

### CODE REVIEW

File: examples/nlp/text2sparql/data/import_datasets.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        logging.info(f"Downloading: {url}")
+        if if_exist(infold, [file]):
+            logging.info("** Download file already exists, skipping download")
+        else:
+            req = Request(url, headers={"User-Agent": "Mozilla/5.0"})
+            with open(os.path.join(infold, file), "wb") as handle:
+                handle.write(urlopen(req, timeout=20).read())
+
+
+def process_text2sparql(infold, outfold, do_lower_case):
```
Comment (jekirl) [inline examples/nlp/text2sparql/data/import_datasets.py:47]: add types

File: examples/nlp/text2sparql/data/import_datasets.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import csv
+import os
+from urllib.request import Request, urlopen
+
+from nemo.collections.nlp.data.data_utils.data_preprocessing import MODE_EXISTS_TMP, if_exist
+from nemo.utils import logging
+
+
+def download_text2sparql(infold):
+    base_url = "http://m.meetkai.com/public_datasets/knowledge/"
```
Comment (jekirl) [inline examples/nlp/text2sparql/data/import_datasets.py:26]: use https

File: examples/nlp/text2sparql/data/import_datasets.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import argparse
+import csv
+import os
+from urllib.request import Request, urlopen
+
+from nemo.collections.nlp.data.data_utils.data_preprocessing import MODE_EXISTS_TMP, if_exist
+from nemo.utils import logging
+
+
+def download_text2sparql(infold):
```
Comment (jekirl) [inline examples/nlp/text2sparql/data/import_datasets.py:25]: specify values of `infold` in doc string

File: examples/nlp/text2sparql/data/import_datasets.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        if if_exist(outfold, [mode_file_name]):
+            logging.info(f"** {MODE_EXISTS_TMP.format(mode, dataset_name, os.path.join(outfold, mode_file_name))}")
+            continue
+
+        if not if_exist(infold, [mode_file_name]):
+            logging.info(f"** {mode} mode of {dataset_name}" f" is skipped as it was not found")
+            continue
+
+        lines = _read_tsv(input_file=os.path.join(infold, mode_file_name))
```
Comment (jekirl) [inline examples/nlp/text2sparql/data/import_datasets.py:77]: instead of loading all of the lines into memory it would make more sense to iterate over the reader here since you are really just streaming them from one file to another with a transform

File: nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        No special modification required for Lightning, define it as you normally would
+        in the `nn.Module` in vanilla PyTorch.
+        """
+        outputs = self.model(
+            input_ids=input_ids, attention_mask=attention_mask, decoder_input_ids=decoder_input_ids, labels=labels,
+        )
+        return outputs
+
+    @typecheck.disable_checks()
+    def generate(self, input_ids):
```
Comment (jekirl) [inline nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py:129]: add types

File: nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                encoder=cfg.language_model.pretrained_encoder_model_name,
+                decoder=cfg.language_model.pretrained_decoder_model_name,
+            )
+
+        self.perplexity_metric = Perplexity()
+
+        self.setup_optimization(cfg.optim)
+
+    @typecheck()
+    def forward(self, input_ids, attention_mask=None, decoder_input_ids=None, labels=None):
```
Comment (jekirl) [inline nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py:118]: add types

File: nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        input_ids, input_mask, decoder_input_ids, labels = batch
+        loss = self.forward(
+            input_ids=input_ids, attention_mask=input_mask, decoder_input_ids=decoder_input_ids, labels=labels
+        )[0]
+
+        tensorboard_logs = {"train_loss": loss, "lr": self._optimizer.param_groups[0]["lr"]}
+
+        return {"loss": loss, "log": tensorboard_logs}
+
+    def validation_step(self, batch, batch_idx):
```
Comment (jekirl) [inline nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py:155]: add types

File: nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+            input_ids=input_ids,
+            pad_token_id=self.encoder_tokenizer.pad_id,
+            bos_token_id=self.encoder_tokenizer.bos_id,
+            eos_token_id=self.encoder_tokenizer.eos_id,
+            decoder_start_token_id=self.decoder_tokenizer.bos_id,
+            **self.cfg.generate,
+        )
+        return outputs
+
+    def training_step(self, batch, batch_idx):
```
Comment (jekirl) [inline nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py:141]: add types

File: nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+            input_ids=input_ids, attention_mask=input_mask, decoder_input_ids=decoder_input_ids, labels=labels
+        )[0]
+
+        perplexity = self.perplexity_metric(loss)
+
+        tensorboard_logs = {"val_loss": loss, "perplexity": perplexity}
+
+        return {"val_loss": loss, "log": tensorboard_logs}
+
+    def validation_epoch_end(self, outputs):
```
Comment (jekirl) [inline nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py:171]: add types

File: nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+    @typecheck.disable_checks()
+    def test_step(self, batch, batch_idx):
+        """Lightning calls this inside the test loop with data from the test dataloader."""
+        input_ids, input_mask, decoder_input_ids, labels = batch
+        sequences = self.generate(input_ids=input_ids)
+        return sequences
+
+    @typecheck.disable_checks()
+    def test_epoch_end(self, outputs):
```
Comment (jekirl) [inline nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py:190]: add types

File: nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        :param outputs: list of individual outputs of each validation step.
+        """
+        avg_loss = torch.stack([x["val_loss"] for x in outputs]).mean()
+        perplexity = torch.stack([x["log"]["perplexity"] for x in outputs]).mean()
+        tensorboard_logs = {"val_loss": avg_loss, "perplexity": perplexity}
+        logging.info(f"evaluation perplexity {perplexity.item()}")
+        return {"val_loss": avg_loss, "log": tensorboard_logs}
+
+    @typecheck.disable_checks()
+    def test_step(self, batch, batch_idx):
```
Comment (jekirl) [inline nemo/collections/nlp/models/neural_machine_translation/neural_machine_translation_model.py:183]: add types

**Review by jekirl — CHANGES_REQUESTED** [review]
Left comments. 
1) Should add types where trivial to do so
2) Simplify the dataset reading process

**Comment by okuchaiev** [comment]
The CI breaks on copyright header check. Can you please change this:
```
# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
# Copyright (c) 2020, MeetKai Inc.  All rights reserved.
```
to this:
```
# Copyright (c) 2020, MeetKai Inc.  All rights reserved.
# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```

**Review by okuchaiev — CHANGES_REQUESTED** [review]
Thank you very much for this PR! Sorry for slow response - we were busy getting 1.0.0beta out of the door.
Overall, the code looks good to me, I just have few mostly cosmetic comments - please see inline.
I was able to run training script and notebook. Regarding notebook - I ran into some issues towards the end.

Here is some comments about notebook specifically:
-------------------------------
print("import_datasets.py already exists') => print("import_datasets.py already exists")
print('"Downloading config file...") => print("Downloading config file...")

before you do "nmt_model = NeuralMachineTranslationModel(config.model, trainer=trainer)" you need to do import:
"from nemo.collections.nlp.models import NeuralMachineTranslationModel"

before you first mention q-value, could you please define what it is?

before you use logging, please  add import first:
"from nemo.utils import logging"

if if_exist(infold, [prefix]):  <---- if_exists was not defined

overall, I could not get past download_entities call - it throws 404 error, can you please have a look?

File: examples/nlp/text2sparql/conf/text2sparql_config.yaml [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+name: &name Text2Sparql
+
+trainer:
+  gpus: 1 # the number of gpus, 0 for CPU, or list with gpu indices
+  num_nodes: 1
+  max_epochs: 2 # the number of training epochs
+  max_steps: null # precedence over max_epochs
+  accumulate_grad_batches: 1 # accumulates grads every k batches
+  precision: 16 # 16 to use AMP
+  amp_level: O1 # O1 or O2 if using AMP
```
[the remark this answers is not in the document]
Comment (vinchg) [reply examples/nlp/text2sparql/conf/text2sparql_config.yaml:12]: Thank you for the feedback! I'm working on making the changes and I should be able to update the PR by tonight.

Regarding the hyperparameters, the lower batch size hasn't been an issue to me and the model performs well on my easy test set. Ideally I would set it higher but my hardware is limited currently.

**Review by okuchaiev — APPROVED** [review]
Thank you!
